core/src/unique.rs

```diff
@@ -263,28 +362,32 @@ impl<T> Future for UniqueConnecFuture<T>
         let mut inner = inner.lock();
         match mem::replace(&mut *inner, UniqueConnecInner::Empty) {
             UniqueConnecInner::Empty => {
-                // This can happen if `set_until()` is called, and the future expires before the
-                // future returned by `get()` gets polled. This means that the connection has been
+                // This can happen if `tie_*()` is called, and the future expires before the
+                // future returned by `dial()` gets polled. This means that the connection has been
                 // closed.
                 Err(IoErrorKind::ConnectionAborted.into())
             },
-            UniqueConnecInner::Pending { mut tasks_waiting, mut dial_fut } => {
+            UniqueConnecInner::Pending { mut tasks_waiting, mut dial_fut, interrupt } => {
                 match dial_fut.poll() {
                     Ok(Async::Ready(())) => {
                         // This happens if we successfully dialed a remote, but the callback
-                        // doesn't call `set_until`. This can be a logic error by the user,
+                        // doesn't call `tie_*`. This can be a logic error by the user,
                         // but could also indicate that the user decided to filter out this
                         // connection for whatever reason.
                         *inner = UniqueConnecInner::Errored(IoErrorKind::ConnectionAborted.into());
                         Err(IoErrorKind::ConnectionAborted.into())
                     },
                     Ok(Async::NotReady) => {
-                        tasks_waiting.push(task::current());
-                        *inner = UniqueConnecInner::Pending { tasks_waiting, dial_fut };
+                        static NEXT_TASK_ID: AtomicUsize = AtomicUsize::new(0);
+                        task_local! {
+                            static TASK_ID: usize = NEXT_TASK_ID.fetch_add(1, Ordering::Relaxed)
+                        }
+                        tasks_waiting.insert(TASK_ID.with(|&k| k), task::current());
+                        *inner = UniqueConnecInner::Pending { tasks_waiting, dial_fut, interrupt };
```

Comment:
why not just keep the `Vec` that we had before and append the tasks but go through the trouble of crating unique `usize` per task and storing them in a hashmap, if we never use them to look up or remove them by that `usize`?